### ISSUE TITLE
Error on clicking the download button twice?

### ISSUE THREAD

**Issue by viveknaskar** [bpampuch/pdfmake#1320]
I have written code such that I am pushing the values one by one.

`vm.pdfInfo.push({text: 'ABC', style: 'abc'});`
 `vm.pdfInfo.push({text: 'Test App', style: 'header'});`
 `vm.pdfInfo.push({text: documentType(), style: 'subheader'});`
`createStudentInformation();`

Here createStudentInformation() does take some dynamic values from the json based on some parameters and pushes like:

`var createStudentInformation = function() {`
                   ` var studentInformation = {};`                   
                   ` $.each(vm.dynamicConfig.assignments.science.fields, function (index, field) {`
                        `if (field.name === 'studentInformation'){`
                           ` studentInformation = field.value;`
                            `mappings = field.mappings;`
                            `formatFunctions = field.formatFunctions;`
                           ` blocks = field.blocks;`
                       ` }`
                    `});`
                   ` if (studentInformation !== {}) {`
                        `$.each(mappings, function (index, mapping) {`
                           `var dispVal = '';`
                          ` var formatFunction = formatFunctions[index];`
                           `if (formatFunction !== '') {`
                             `  dispVal = $scope[formatFunction](dispVal);`
                           `}`
                          ` if (mapping !== '') {`
                              `studentInformation.table.body[index+1].push(dispVal);`
                           `}`
                       ` });`
                        `vm.pdfInfo.push(studentInformation);`
                    `}`
               ` };`

I have a requirement such that I have to use methods like above and push into the document. The values for the table is coming from the json. I am able to download the pdf for the first time, but when I am clicking the download button again, I am getting an error like below -

pdfmake.js:1743 Uncaught TypeError: Cannot read property '_calcWidth' of undefined
    at pdfmake.js:1743
    at n.auto (pdfmake.js:1743)
    at i.processRow (pdfmake.js:1743)
    at i.processTable (pdfmake.js:1743)
    at pdfmake.js:1743
    at pdfmake.js:1743
    at i.processNode (pdfmake.js:1743)
    at pdfmake.js:1743
    at Array.forEach (<anonymous>)
    at i.processVerticalContainer (pdfmake.js:1743)

**Comment by liborm85** [comment]
You can not call it twice with the same `docDefinition` data. `docDefinition` needs generated again on clicking.

**Comment by viveknaskar** [comment]
I have found out what the problem was. Since I am using the global object from the JSON, i.e., vm.dynamicConfig.assignments, the object is used again and again. 

We can fix this by cloning the javascript object using JSON.parse(JSON.stringify(obj))

I have made the following changes:
var assignments = JSON.parse(JSON.stringify(vm.dynamicConfig.assignments));

I updated the object for iterating over my fields:

$.each(assignments.science.fields, function (index, field) {
.
.
.
}

It works like charm.

Thanks @liborm85 for the comment, because of whom I got the idea.
